src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs

```diff
@@ -84,8 +85,12 @@ private static (IReadOnlyDictionary<FieldDeclaration, InputModelProperty> Fields
 
             foreach (var inputModelProperty in inputModel.Properties)
             {
+                // property is read-only if:
+                // * model is read-only
+                // * property is a collection
+                // * property is required and model is either input only or output only
                 var fieldModifiers = inputModelProperty.IsReadOnly || inputModelProperty.Type is InputDictionaryType or InputListType ||
-                    (inputModel.Usage == InputModelTypeUsage.Input && inputModelProperty.IsRequired)
+                    ((inputModel.Usage == InputModelTypeUsage.Input || inputModel.Usage == InputModelTypeUsage.Output) && inputModelProperty.IsRequired)
```

Comment:
`inputModel.Usage is InputModelTypeUsage.Input or InputModelTypeUsage.Output` is a bit shorter